### PR TITLE
add constant folding

### PR DESCRIPTION
**Pull request by XuHuaiyu**
PTAL
@shenli @hanfei19910905

### CODE REVIEW

**Comment by hanfei19910905** [comment]
@coocood @zimulala @tiancaiamao PTAL

File: expression/expression.go [hunk 1]
```diff
@@ -217,17 +217,43 @@ func (sf *ScalarFunction) ToString() string {
...
 		return nil, nil
 	}
+
 	if len(args) < f.MinArgs || (f.MaxArgs != -1 && len(args) > f.MaxArgs) {
 		return nil, evaluator.ErrInvalidOperation.Gen("number of function arguments must in [%d, %d].",
 			f.MinArgs, f.MaxArgs)
 	}
+
+	canConstantFolding := true
+	var datums []types.Datum
```
Comment (hanfei19910905) [inline expression/expression.go]: datums := make([]types.Datum, 0, len(args)) , this can avoid slice growing.

File: plan/logical_plan_builder.go [hunk 2]
```diff
@@ -141,7 +141,8 @@ func extractOnCondition(conditions []expression.Expression, left LogicalPlan, ri
 					continue
 				}
 				if left.GetSchema().GetIndex(rn) != -1 && right.GetSchema().GetIndex(ln) != -1 {
-					cond, _ := expression.NewFunction(ast.EQ, types.NewFieldType(mysql.TypeTiny), rn, ln)
+					tmpCond, _ := expression.NewFunction(ast.EQ, types.NewFieldType(mysql.TypeTiny), rn, ln)
+					cond, _ := tmpCond.(*expression.ScalarFunction)
 					eqCond = append(eqCond, cond)
```
Comment (hanfei19910905) [inline plan/logical_plan_builder.go]: eqCond = append(eqCond, tmpCond.(*expression.ScalarFunction))

File: expression/expression.go [hunk 3]
```diff
@@ -217,17 +217,43 @@ func (sf *ScalarFunction) ToString() string {
...
+		fn := f.F
+		newArgs, err := fn(datums, nil)
+
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return &Constant{
+			Value:   newArgs,
+			RetType: retType,
+		}, errors.Trace(err)
```
Comment (hanfei19910905) [inline expression/expression.go]: the second returning value can be nil

Comment (XuHuaiyu) [inline expression/expression.go]: ok, thank you !

File: expression/expression.go [hunk 4]
```diff
@@ -217,17 +217,43 @@ func (sf *ScalarFunction) ToString() string {
...
+
 	if len(args) < f.MinArgs || (f.MaxArgs != -1 && len(args) > f.MaxArgs) {
 		return nil, evaluator.ErrInvalidOperation.Gen("number of function arguments must in [%d, %d].",
 			f.MinArgs, f.MaxArgs)
 	}
+
+	canConstantFolding := true
+	datums := make([]types.Datum, 0, len(args))
+
+	for _, expr := range args {
```
Comment (shenli) [inline expression/expression.go]: rand() has no args, but it should not be folded.
mysql> select *, rand() from t1;
+------+------+--------------------+
| c1   | c2   | rand()             |
+------+------+--------------------+
|    2 |    3 | 0.8671364708497529 |
|    4 |    4 | 0.1637872002681598 |
+------+------+--------------------+

File: expression/expression.go [hunk 5]
```diff
@@ -217,17 +217,43 @@ func (sf *ScalarFunction) ToString() string {
...
 }
 
-// NewFunction creates a new scalar function.
-func NewFunction(funcName string, retType *types.FieldType, args ...Expression) (*ScalarFunction, error) {
+// NewFunction creates a new scalar function or constant.
+func NewFunction(funcName string, retType *types.FieldType, args ...Expression) (Expression, error) {
 	f, ok := evaluator.Funcs[funcName]
 	if !ok {
 		log.Errorf("Function %s is not implemented.", funcName)
 		return nil, nil
```
Comment (tiancaiamao) [inline expression/expression.go]: return error if it's an error...

File: plan/logical_plan_builder.go [hunk 6]
```diff
@@ -142,7 +142,7 @@ func extractOnCondition(conditions []expression.Expression, left LogicalPlan, ri
 				}
 				if left.GetSchema().GetIndex(rn) != -1 && right.GetSchema().GetIndex(ln) != -1 {
 					cond, _ := expression.NewFunction(ast.EQ, types.NewFieldType(mysql.TypeTiny), rn, ln)
-					eqCond = append(eqCond, cond)
+					eqCond = append(eqCond, cond.(*expression.ScalarFunction))
```
Comment (tiancaiamao) [inline plan/logical_plan_builder.go]: make sure this operation won't crash the whole program

Comment (XuHuaiyu) [inline plan/logical_plan_builder.go]: variable rn and ln are both Column type
so EQ would not be constant folded, cond can be promised to be ScalarFunction.

File: evaluator/builtin.go [hunk 7]
```diff
@@ -167,6 +167,24 @@ var Funcs = map[string]Func{
 	ast.GetVar:     {builtinGetVar, 1, 1},
 }
 
+// FuncsCnntBeCnstntFld is for functions that
+// use input parameter ctx or
+// return an uncertain result would not be constant folded
+// the value 0 means nothing
+var FuncsCnntBeCnstntFld = map[string]int{
```
Comment (shenli) [inline evaluator/builtin.go]: DynamicFuncs may be a better name. And this variable is only used in this package, so we can make it private.

Comment (shenli) [inline evaluator/builtin.go]: And add a few test cases for these functions.

Comment (XuHuaiyu) [inline evaluator/builtin.go]: ok

**Comment by hanfei19910905** [comment]
LGTM

File: evaluator/builtin.go (hunk 7 above)
Comment (XuHuaiyu) [inline evaluator/builtin.go]: I did some fix, and DynamicFuns would be used in package 'expression', so I make it public

**Comment by shenli** [comment]
LGTM
Well done!

**Pull request merged**